pkg/kubelet/kubelet.go

```diff
@@ -1750,24 +1750,26 @@ func (kl *Kubelet) syncLoop(updates <-chan kubetypes.PodUpdate, handler SyncHand
 	defer housekeepingTicker.Stop()
 	plegCh := kl.pleg.Watch()
 
-	// If the starting systemd unit asks for a watchdog, report keep-alive periodically
-	var watchdogCh <-chan time.Time
-	if watchdogDuration, err := daemon.SdWatchdogEnabled(false); watchdogDuration != 0 || err != nil {
+	// If the starting systemd unit asks for a watchdog, notify periodically as long as the sync loop is called
+	if watchdogTimeout, err := daemon.SdWatchdogEnabled(false); watchdogTimeout != 0 || err != nil {
 		switch {
 		case err != nil:
-			klog.Errorf("Checking watchdog configuration failed: %v", err)
-			watchdogCh = make(chan time.Time)
+			klog.Errorf("Reading watchdog configuration failed: %v", err)
+			break
+		case watchdogTimeout != 0:
+			notifyPeriod := watchdogTimeout / 3
+			loopTimeout := 2 * kl.resyncInterval
+			klog.Infof("Arming watchdog: configured watchdog timeout is %v, will report every %v, and stop reporting when sync loop is stalled for more than %v", watchdogTimeout, notifyPeriod, loopTimeout)
+			go wait.Forever(func() {
+				// check syncLoopMonitor: if it reports an old time, something is going wrong
+				if kl.clock.Now().After(kl.LatestLoopEntryTime().Add(loopTimeout)) {
+					klog.Warningf("SyncLoopMonitor not updated for at least %v, not going to notify watchdog", loopTimeout)
```

Comment:
Sorry to quibble, but I want to know: could this end up spamming the logs with many reports of a stall?

Reply:
Yes, fair point, fixed.
And rebased...